examples/petsc/ex28p.cpp

```diff
@@ -0,0 +1,360 @@
+   e_r(1) = -pow(k0*2.0,2);
+   PWConstCoefficient e_r_func(e_r);
+
+   // 9. Define the parallel grid function and parallel linear forms.
+   BlockVector x(block_offsets);
+   BlockVector trueX(block_trueOffsets);
+
+   // Define the boundary attributes
+   Array<int> ess_bdr;
+   if (pmesh->bdr_attributes.Size())
+   {
+      ess_bdr.SetSize(pmesh->bdr_attributes.Max());
+      ess_bdr = 0;
+   }
+
+   // 10. Assemble the finite element matrices for the LHS and RHS
+   //
+   //                            A = [ Att  0 ]
+   //                                [ 0    0 ]
+   //
+   //                            B = [ Btt Bzt ]
+   //                                [ Btz Bzz ]
+   ParBilinearForm *att = new ParBilinearForm(N_space);
+   ParBilinearForm *btt = new ParBilinearForm(N_space);
+   ParBilinearForm *azz = new ParBilinearForm(L_space);
+   ParBilinearForm *bzz = new ParBilinearForm(L_space);
+   ParMixedBilinearForm *btz = new ParMixedBilinearForm(N_space, L_space);
+
+   PetscParMatrix *pAtt = NULL, *pBtt = NULL, *pBzz = NULL;
+   PetscParMatrix *pAzz = NULL, *pBtz = NULL, *pBzt = NULL;
+   Operator::Type tid = Operator::PETSC_MATAIJ;
+   OperatorHandle Atth(tid), Btth(tid), Bzzh(tid), Btzh(tid), Azzh(tid);
+
+   att->AddDomainIntegrator(new CurlCurlIntegrator(u_r_func));
+   att->AddDomainIntegrator(new VectorFEMassIntegrator(e_r_func));
+   att->Assemble();
+   att->EliminateEssentialBCDiag(ess_bdr, 1.0);
+   att->Finalize();
+   att->ParallelAssemble(Atth);
+   Atth.Get(pAtt);
+   Atth.SetOperatorOwner(false);
+
+   // A dummy Azz is required to set the block size and apply the essential
+   // boundary condition
+   azz->Assemble();
+   azz->EliminateEssentialBCDiag(ess_bdr, 1.0);
+   azz->Finalize();
+   azz->ParallelAssemble(Azzh);
+   Azzh.Get(pAzz);
+   Azzh.SetOperatorOwner(false);
+
+   btt->AddDomainIntegrator(new VectorFEMassIntegrator(u_r_func));
+   btt->Assemble();
+   btt->EliminateEssentialBCDiag(ess_bdr, numeric_limits<double>::min());
+   btt->Finalize();
+   btt->ParallelAssemble(Btth);
+   Btth.Get(pBtt);
+   Btth.SetOperatorOwner(false);
+   (*pBtt) *= -1;
+
+   bzz->AddDomainIntegrator(new DiffusionIntegrator(u_r_func));
+   bzz->AddDomainIntegrator(new MassIntegrator(e_r_func));
+   bzz->Assemble();
+   btt->EliminateEssentialBCDiag(ess_bdr, numeric_limits<double>::min());
+   bzz->Finalize();
+   bzz->ParallelAssemble(Bzzh);
+   Bzzh.Get(pBzz);
+   Bzzh.SetOperatorOwner(false);
+   (*pBzz) *= -1;
+
+   ParLinearForm dummy(N_space);
+   btz->AddDomainIntegrator(new MixedVectorWeakDivergenceIntegrator(u_r_func));
+   btz->Assemble();
+   btz->EliminateTestDofs(ess_bdr);
+   btz->EliminateTrialDofs(ess_bdr,x.GetBlock(0),dummy);
+   btz->Finalize();
+   btz->ParallelAssemble(Btzh);
+   Btzh.Get(pBtz);
+   Btzh.SetOperatorOwner(false);
+
+   pBzt = pBtz->Transpose();
+
+   PetscParMatrix *LHSOp = NULL, *RHSOp = NULL;
+   // We construct the BlockOperator and we then convert it to a
```

Comment:
Are you using a MATNEST here with SLEPc?. If so, you should mention it here that you are preserving the blocked nature of the operator

Reply:
Yes this creates a MATNEST so only some solvers are available (e.g. Jacobi-Davidson method with Jacobi preconditioner). I can add a variation of the problem where the matrix is converted to MATAIJ and we can use e.g. shift-and-invert with a direct solver. I will improve the comments.

Reply:
It would be nice to show users the flexibility of the PETSc matrix formats. If you add this, please also add a comment that converting from NEST to AIJ in sequential is reasonably fast, while doing in parallel may be a bit slower (In PETSc we currently do no handle fast conversion when the blocks are MPIAIJ; this could be added in the future if needed)

Reply:
You are correct, I also observed that the conversion takes longer in parallel. However, for my specific problem, the shift-and-invert method with a direct solver seems faster than using the Jacobi-Davidson method. Do you think using `HypreParMatrixFromBlocks` and then converting to a `PetscParMatrix` would be faster (potentially using a Hypre preconditioner in PETSc)?

Reply:
If it is specifically written for mfem block operator, the answer is yes. However, my personal opinion is that it is best to keep the example as simple as possible. If you want users experiments with shift-and-invert, just add the conversion operator from the PetscParMatrix class (with the caveat about performances)

Reply:
Yes I should keep the example simple. Please see the latest version: https://github.com/mfem/mfem/blob/1855ec2993892ba4e5835bb3977278a93bfae852/examples/petsc/ex28p.cpp#L261-L274